### DISCUSSION TITLE
sqlite3 在處理bcrpt 時遇到的問題

### DISCUSSION THREAD

**Discussion by pocper1** [1101-GS3073-C/Python-Community-GS3073-C#43] [Q&A] [answered]
這是遇到的問題
![image](https://user-images.githubusercontent.com/76590438/136657080-32456b23-8341-4dcc-82f7-800c09c97883.png)

我照著網路上先import flask_bcrpt，
![image](https://user-images.githubusercontent.com/76590438/136657113-a25ef681-8811-4a91-ae12-32ad44bac5a1.png)

之後再使用function
![image](https://user-images.githubusercontent.com/76590438/136657127-9ac31ea9-45ae-4660-9918-fff25503b1ec.png)
就遇到上面的問題了，我也不清楚哪裡有做錯
(hash_password 我有印出來 是有值的，但是好像在insert into database 遇到問題)

**Comment by edaw1111** [comment]
他有跳出error之類的嗎？ 還是下面那張圖就是他跳出來的error？

**Reply by pocper1** [reply]
第一張圖就是error

**Reply by edaw1111** [reply]
你可以截圖整個程式碼嗎? 我這樣真的看不太出來問題在哪邊。

**Reply by pocper1** [reply]
註冊畫面
![image](https://user-images.githubusercontent.com/76590438/136785299-b19dbed1-2088-4e86-b826-e74b70aeae98.png)

程式碼的部分
![image](https://user-images.githubusercontent.com/76590438/136785380-03e9f20a-f42e-423f-bcb9-f7153b98ce25.png)
![image](https://user-images.githubusercontent.com/76590438/136785399-1d24870f-ba2a-47ff-986d-a90c38cd5b21.png)

好像是經過bcrypt.generate_password_hash(password1) function 出來 就會有問題

**Comment by edaw1111** [comment, marked as answer]
問題並不是出在bcrypt.generate_password_hash(password1)這行程式碼喔，應該是在下面那行 cur.execute(f"INSERT INTO user (`username`, `email`, `password`) VALUES ('{username}','{email}','{hash_password}')")，一般來說在sql語法裡VALUES裡面放的值會希望是string字串的型態，但是你經過bcrypt.generate_password_hash(password1)出來的 hash_password，如果你有看過他print出來的值的話，你會發現前面多了一個b，這是另一種byte格式的字串，和string字串的型態不太相同，這部分你可以參考一下這篇[文章](https://dotblogs.com.tw/Eyelash/2020/05/26/174740)，所以當你把hash_password放入VALUES 裡的時候就會出現syntax error，意思是語法的錯誤。

**Reply by pocper1** [reply]
ok 理解了! 謝謝~
